lib/tests/test_fqdnauth.py

```diff
@@ -0,0 +1,188 @@
+        target_ip_resolved=['badhost1', 'host1234', 'badhost2'],
+        whitelist_in_config=['host4567', 'foo', 'host1234', 'bar'],
+        expected_verdict='ACCEPT'
+    )
+
+
+def test_space_comma_and_newline_delimiters_all_work_in_whitelist():
+    testconfig = dedent("""
+        [plugin]
+        [hosts_for_groups]
+        __all__ = host0, host1 host2
+                  host3
+    """)
+
+    with patch('safeguard.sessions.plugin.host_resolver.HostResolver.resolve_hosts_by_ip') as mock_resolver:
+        for i in range(4):
+            mock_resolver.return_value = ['host' + str(i)]
+            result = Plugin(configuration=testconfig).authorize({},{},target_server='DONTCARE', target_username='test_user')
+            assert result['verdict'] == 'ACCEPT'
+
+
+def test_matching_falls_back_to_ip_match_if_resolving_failed():
+    _assert_for_verdict(
+        target_ip='1.2.3.4',
+        target_ip_resolved=[],
+        whitelist_in_config=['host4567', 'foo', 'host1234', '1.2.3.4'],
+        expected_verdict='ACCEPT'
+    )
+
+
+def test_session_accepted_if_match_found_for_gateway_group():
+    testconfig = dedent("""
+        [plugin]
+        [hosts_for_groups]
+        test_group = host1234, other_host, foo, bar
+    """)
+
+    with patch('safeguard.sessions.plugin.host_resolver.HostResolver.resolve_hosts_by_ip') as mock_resolver:
+        mock_resolver.return_value = ['host1234']
+        result = Plugin(configuration=testconfig).authorize({},{},
+                    target_server='DONTCARE',
+                    gateway_groups=['fake_group', 'foo', 'bar', 'test_group'],
+                    target_username='test_user'
+                )
+        assert result['verdict'] == 'ACCEPT'
+
+
+def test_wildcards_can_be_used_in_host_lists_for_single_resolved_hostname():
+    _assert_for_verdict(
+        target_ip='DONTCARE',
+        target_ip_resolved=['host1234'],
+        whitelist_in_config=['h?st*'],
+        expected_verdict='ACCEPT'
+    )
+
+
+def test_wildcards_can_be_used_in_host_lists_for_multiple_resolved_hostname():
+    _assert_for_verdict(
+        target_ip='DONTCARE',
+        target_ip_resolved=['host1234', 'foobar'],
+        whitelist_in_config=['*bar'],
+        expected_verdict='ACCEPT'
+    )
+
+
+def test_groups_for_host_specification_works_for_single_group():
+    testconfig = dedent("""
+        [plugin]
+        [groups_for_hosts]
+        host1234 = test_group
+    """)
+    with patch('safeguard.sessions.plugin.host_resolver.HostResolver.resolve_hosts_by_ip') as mock_resolver:
+        mock_resolver.return_value = ['host1234']
+        result = Plugin(configuration=testconfig).authorize({},{},
+                    target_server='DONTCARE',
+                    gateway_groups=['fake_group', 'test_group'],
+                    target_username='test_user'
+                )
+        assert result['verdict'] == 'ACCEPT'
+
+
+def test_groups_for_host_specification_supports_wildcards_for_hosts():
```

Comment:
I'm missing a case where groups_for_hosts doesn't match and we should get DENY - as far as I understand.